integration-test/src/main/java/com/sequenceiq/it/cloudbreak/context/TestContext.java

```diff
@@ -186,6 +186,64 @@ public InstanceAwait getInstanceAwait() {
         return instanceAwait;
     }
 
+    public <E extends Exception, T extends CloudbreakTestDto, U extends MicroserviceClient> T whenException(Class<T> entityClass,
+            Class<? extends MicroserviceClient> clientClass, Action<T, U> action, Class<E> expectedException) {
+        return whenException(entityClass, clientClass, action, expectedException, emptyRunningParameter());
+    }
+
+    public <E extends Exception, T extends CloudbreakTestDto, U extends MicroserviceClient> T whenException(Class<T> entityClass,
+            Class<? extends MicroserviceClient> clientClass, Action<T, U> action, Class<E> expectedException, RunningParameter runningParameter) {
+        return whenException(getEntityFromEntityClass(entityClass, runningParameter), clientClass, action, expectedException, runningParameter);
+    }
+
+    public <E extends Exception, T extends CloudbreakTestDto, U extends MicroserviceClient> T whenException(T entity,
+            Class<? extends MicroserviceClient> clientClass, Action<T, U> action, Class<E> expectedException) {
+        return whenException(entity, clientClass, action, expectedException, emptyRunningParameter());
+    }
+
+    public <E extends Exception, T extends CloudbreakTestDto, U extends MicroserviceClient> T whenException(T entity,
+            Class<? extends MicroserviceClient> clientClass, Action<T, U> action, Class<E> expectedException, RunningParameter runningParameter) {
+        checkShutdown();
+        String key = runningParameter.getKey();
+        if (key == null || key.isEmpty()) {
+            key = action.getClass().getSimpleName();
+        }
+        if (!getExceptionMap().isEmpty() && runningParameter.isSkipOnFail()) {
+            LOGGER.info("Should be skipped because of previous error. when [{}]", key);
+            return entity;
+        }
+
+        CloudbreakUser who = getWho(runningParameter);
+
+        LOGGER.info("when {} action on {} by {}, name: {}", key, entity, who, entity.getName());
+        Log.when(LOGGER, action.getClass().getSimpleName() + " action on " + entity + " by " + who);
+
+        try {
+            return doAction(entity, clientClass, action, who.getAccessKey());
+        } catch (Exception e) {
+            if (!e.getClass().equals(expectedException)) {
+                String message = String.format("Expected exception (%s) does not match with the actual exception (%s).",
+                        expectedException, e.getClass());
+                getExceptionMap().put("when", new TestFailException(message));
```

Comment:
Mybe do you want to write the `whenException` in the log?

Reply:
This is intentional. If the actual and expected exception don't met, we should see a `TestFailException` and put this to the `ExceptionMap` for later validation. Along with this we will see `the message` in the console log and the formatted HTML log as well.

Reply:
I can easier to find the method by action. If I read the `whenException`, I know where should I search in the code. If I read `when` probably I will search the when methods

Reply:
That's the point! You are right, I've overlooked your comment in the code.